nixos/netmaker: init alongside it's dependencies

nazarewk opened the pull request:
## Description of changes

Part of an effort to run SSO-enabled Netmaker on NixOS:
- https://github.com/gravitl/netmaker/pull/2785
- https://github.com/gravitl/netclient/pull/682
- https://github.com/gravitl/netmaker-ui-2/pull/431
- https://github.com/NixOS/nixpkgs/pull/283768

This uses my own fork of netmaker because I was blocked by too many errors on every single original repository involved when trying to get SSO to work at all.

The main idea is to get something fully working merged to nixpkgs as soon as possible and then refine it in small patches (including moving to upstream after merge of fork).

NOTE: this will not allow your peers to talk to each other from behind NAT, see https://github.com/NixOS/nixpkgs/pull/283768#issuecomment-1918805435

## Things done

Packaged up and configured all 3 dependencies (+ alternate unfree server version).

Tested through https://github.com/nazarewk-iac/nix-configs/commit/c48a7242bff6fa4d405b7624a59b6e553574d06e

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

qjoly reviewed (approved):
Great work ! Thank you

wexder commented on pkgs/by-name/ne/netmaker-ui/package.nix:
```diff
@@ -0,0 +1,33 @@
+  version = "unstable-2024-01-22";
+
+  src = fetchFromGitHub {
+    owner = "nazarewk";
+    repo = "netmaker-ui-2";
```
Why are we using fork and not the official repository ?

nazarewk replied:
added note in a description

wexder commented on pkgs/by-name/ne/netmaker/package.nix:
```diff
@@ -0,0 +1,51 @@
+  passthru.updateScript = nix-update-script { extraArgs = [ "--version=branch=develop" ]; };
+  version = "unstable-2024-01-24";
+
+  src = fetchFromGitHub {
+    owner = "nazarewk";
```
Again, why not official repository ?

nazarewk replied:
added note in a description

wexder replied:
I've noticed, sot after your patches are merged into official repository we will be switching back to official repository?
I don't know what's best practice in nixpkgs about things like this

nazarewk replied:
yeah, this is the idea

wexder replied:
I'm really not sure what's the best practice about this within the nixpkgs, so we will need to discuss with someone more involved.

wexder commented on pkgs/by-name/ne/netmaker/package.nix:
```diff
@@ -0,0 +1,51 @@
+    description = "WireGuard automation from homelab to enterprise - Community Edition";
+    homepage = "https://netmaker.io";
+    changelog = "https://github.com/gravitl/netmaker/-/releases/v${version}";
+    license = lib.licenses.asl20;
```
Pro version of netmaker is licensed under `The Netmaker Enterprise license` not `Apache-2.0`

nazarewk replied:
this one is not a pro version

wexder replied:
True I'm sorry , my oversight

SuperSandro2000 replied:
Then we should also drop the isPro argument

nazarewk replied:
That `isPro` is today's addition, I am building 3 flavors of the package through overrides:
- `netmaker` - has only the server binary (`.` subpackage)
- `nmctl` - has only `nmctl` binary (`cli` subpackage)
- `netmaker-pro` - has only the server binary and is licensed as `unfree` (`/pro` subdirectory is not used in `nmctl` at all)

wexder commented on nixos/modules/services/networking/netmaker/netclient.nix:
```diff
@@ -0,0 +1,75 @@
+      description = lib.mdDoc ''
+        Trusts all traffic coming from Netmaker interface.
+
+        To disable you can set up rules with
```
Seems like description is cut off ?

wexder commented on nixos/modules/services/networking/netmaker/netmaker.nix:
There's quite few services enabled together with the netmaker server. Aren't these going to clash with services that user would have configured in their configuration ?

wexder replied:
Please keep in mind I'm not nixos expert. Just the use of multiple services inside one like this seems like not practice.

nazarewk replied:
Yeah I also thought about it, but wanted to push out something that works before refining it indefinitely:
- `mosquitto` could clash on:
	- `listener 9001` which could be adjusted by changing `mq.internal.port` or `internalListenAddress`,
	- topic names: would require prefixing on netmaker code side (mayble I'll do it)
- `coredns` seems to only be configured to read `netmaker.hosts` file:
	- I don't really know how to compose it properly in nixos config so it doesn't clash with other CoreDNS config
- `caddy` should be fine as the configuration is `virtualHosts` scoped except:
	```
      networking.firewall.allowedTCPPorts = [ 80 443 ];
      systemd.services.caddy.serviceConfig.AmbientCapabilities = [ "CAP_NET_BIND_SERVICE" ];
	```

wexder replied:
- I think there's potential issue with `mosquitto` in case of `services.mosquitto.listeners`, where we should merge the potential objects in user configuration and the one we are defining here to allow additional listeners to be defined by user. 
- `coredns` is probably be harder to correctly merge user defined config and one required for `netmaker`
- In caddy I've missed the `networking.firewall.allowedTCPPorts = [ 80 443 ];` setting, this shouldn't be enabled by default and should be configurable. I don't know what does the `AmbientCapabilities` and how to handle them.

nazarewk replied:
- I'm pretty sure `mosquitto.listeners` list gets merged by the nix module so it should not be an issue.
- `coredns` is covered by [reusable snippet](https://coredns.io/manual/configuration/#reusable-snippets) mixed with `coredns.defaults.config`
- covered by `cfg.webserver.openFirewall` defaulting to `true`
- I'm pretty sure that `CAP_NET_BIND_SERVICE` is required for caddy to listen on low ports (80/443) because it runs at it's own user

generally Netmaker wants to be exposed to the internet on a specific domain with TLS, the more I dig into it the more I see it could accommodate to non-public TLS running on private network itself (could probably implement nixos test in containers), but let's not get ahead of ourselves and implement everything at once.

wexder replied:
Is it required for caddy to run on 80/443 ? Let's say I would rather run it on different ports, then I shouldn't need the `CAP_NET_BIND_SERVICE` correct ?
80/443 is good default but we allowing users to change seems like a good option. But maybe it's too much options

nazarewk replied:
I don't think anything except 80/443 would work for ACME certificate issuance at which point you'd be better off turning off Caddy integration and setting this up yourself.

wexder replied:
Ah you are correct

wexder commented on pkgs/by-name/ne/netclient/package.nix:
```diff
@@ -1,34 +1,62 @@
-
+  subPackages = [
+    "."
+    # "gui" # should be packaged separately because it's another binary with many more dependencies
```
Should this be left commented out here ?

nazarewk replied:
yes, in case somebody wondered why aren't we building GUI part.

I tried to quickly build it, but it didn't work at all so left it out completely.

wexder replied:
ok make sense

wexder commented on nixos/modules/services/networking/netmaker/netmaker.nix:
```diff
@@ -0,0 +1,625 @@
+            if !cfg.mqtt.acl.restricted then [
+              "readwrite #"
+            ] else [
+              # search for `client.Subscribe` in:
+              # - https://github.com/gravitl/netmaker
+              # - https://github.com/gravitl/netclient
+              # https://github.com/gravitl/netclient
```
Why are these commented out ?

nazarewk replied:
those were supposed to be instructions on finding out set of topics used by Netmaker

wexder commented:
> > @nazarewk also keep in mind that there's open issue with netclient dns feature not working properly #260817
> 
> Yeah, this one was moved back to using CoreDNS afaik, I did not get to the part of testing it. I just know CoreDNS itself returns proper result (using `dig XX @IP`)

The issue is on client side and I don't think coreDNS is being deployed with netclient on client device. But this is just FYI and not subject of this PR

nazarewk commented:
FYI: I successfully wired up CoreDNS to the public DNS network resolving Netmaker's private instances without any clients-side configuration:
```
kdn@krul ~/d/g/n/nix-configs (main)> dig +short wg-0.temporary-network.net.subsidize-stroller.kdn.im.
10.177.68.1
```
seems to fully address https://github.com/NixOS/nixpkgs/issues/260817

wexder commented:
> FYI: I successfully wired up CoreDNS to the public DNS network resolving Netmaker's private instances without any clients-side configuration:
> 
> ```
> kdn@krul ~/d/g/n/nix-configs (main)> dig +short wg-0.temporary-network.net.subsidize-stroller.kdn.im.
> 10.177.68.1
> ```
> 
> seems to fully address #260817 

This works if you don't mind having your private dns exposed to the internet. It's  debatable if it's a security risk, but some people might not want to do it.
Anyway cool it works and thanks for all the work, I will probably try to deploy this myself over the weekend

nazarewk commented:
> > FYI: I successfully wired up CoreDNS to the public DNS network resolving Netmaker's private instances without any clients-side configuration:
> > ```
> > kdn@krul ~/d/g/n/nix-configs (main)> dig +short wg-0.temporary-network.net.subsidize-stroller.kdn.im.
> > 10.177.68.1
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > seems to fully address #260817
> 
> This works if you don't mind having your private dns exposed to the internet. It's debatable if it's a security risk, but some people might not want to do it. Anyway cool it works and thanks for all the work, I will probably try to deploy this myself over the weekend

Yeah, this is why it won't work unless you wire it up in upper zone and even if you do it is a default that you can set to `[]` to turn off.

AndersonTorres reviewed (changes requested):
First things first:

**One commit per package**, as described in the contributing guidelines.

After that I can complain about the code itself.

wexder commented:
@nazarewk For some reason the `netmaker-generate-config` always fails with error `tr: write error: Broken pipe` when running after startup and has to be restarted manually

nazarewk committed:
netmaker-ui: init at unstable-2024-01-26

Uses unstable because the SSO feature wouldn't work without it.
Can be switched after https://github.com/gravitl/netmaker-ui-2/commit/143ea10fd913cc01fcc9c82b088a37dccdaff438 lands in a release

nazarewk committed:
netclient: 0.22.0 -> unstable-2024-01-26

Uses the fork because the SSO feature wouldn't work without it.
Can be switched to upstream after following PR lands in a release:
 https://github.com/gravitl/netclient/pull/682

nazarewk committed:
netmaker: 0.21.2 -> unstable-2024-01-26

Uses the fork because the SSO feature wouldn't work without it.
Can be switched to upstream after following PR lands in a release:
 https://github.com/gravitl/netclient/pull/682

nazarewk committed:
nmctl: init from netmaker

nazarewk commented:
marking temporarily as a draft ad I have to test whether it works after splitting up into commits.

nazarewk committed:
netmaker-pro: init from netmaker

SuperSandro2000 commented on pkgs/by-name/ne/netmaker-pro/package.nix:
```diff
@@ -0,0 +1 @@
+{ callPackage }: callPackage ../netmaker/package.nix { isPro = true; }
```
This should be in all-packages.nix

SuperSandro2000 commented on pkgs/by-name/ne/netmaker-ui/package.nix:
```diff
@@ -0,0 +1,35 @@
+, buildNpmPackage
+, fetchFromGitHub
+, nix-update-script
+, ...
```
```suggestion
```

SuperSandro2000 commented on pkgs/by-name/ne/netmaker-ui/package.nix:
```diff
@@ -0,0 +1,35 @@
+  pname = "netmaker-ui";
+  version = "unstable-2024-01-26";
+
+  passthru.updateScript = nix-update-script {
```
passthru should be just before meta

SuperSandro2000 commented on pkgs/by-name/ne/netmaker/package.nix:
```diff
@@ -0,0 +1,64 @@
+, withServer ? true
+, withCLI ? false
+, isPro ? false
+, ...
```
```suggestion
```

SuperSandro2000 commented on pkgs/by-name/ne/netmaker/package.nix:
```diff
@@ -0,0 +1,64 @@
+, ...
+}:
+buildGoModule rec {
+  pname = builtins.concatStringsSep "-" (
```
```suggestion
  pname = lib.concatStringsSep "-" (
```
if there is a lib variant of a builtin that is always prefered.

SuperSandro2000 commented on pkgs/by-name/ne/netmaker/package.nix:
```diff
@@ -0,0 +1,64 @@
+  # TODO: switch to upstream after https://github.com/gravitl/netmaker/pull/2785
+  passthru.updateScript = nix-update-script { extraArgs = [ "--version=branch=fix-sso" ]; };
+  src = fetchFromGitHub {
+    #owner = "gravitl";
```
```suggestion
```

SuperSandro2000 commented on pkgs/by-name/ne/netmaker/package.nix:
```diff
@@ -0,0 +1,64 @@
+  version = "unstable-2024-01-26";
+
+  # TODO: switch to upstream after https://github.com/gravitl/netmaker/pull/2785
+  passthru.updateScript = nix-update-script { extraArgs = [ "--version=branch=fix-sso" ]; };
```
passthru should be before meta

SuperSandro2000 commented on pkgs/by-name/ne/netmaker/package.nix:
```diff
@@ -0,0 +1,64 @@
+    repo = "netmaker";
+    rev = "36ffc395eaff2c6e66fc6e17034329292541ca52";
+    hash = "sha256-P19fGDMFKjiR5LP1kQitTVMLKYiPs0z3ICZaIFfgRzQ=";
+    postFetch = lib.optionalString (!isPro) ''
```
We cannot do it like this as it changes the hash one line above.

SuperSandro2000 commented on pkgs/by-name/ne/netmaker/package.nix:
```diff
@@ -0,0 +1,64 @@
+  };
+  vendorHash = "sha256-t7g6Tozq/QLq0/5bpXNDCJrOPTjMlvcDUaD6EGqII3Y=";
+
+  subPackages =
+    lib.optional withServer "." ++
+    lib.optional withCLI "cli"
+  ;
```
```suggestion
  subPackages = lib.optional withServer "."
    ++ lib.optional withCLI "cli";
```

SuperSandro2000 commented on pkgs/by-name/nm/nmctl/package.nix:
```diff
@@ -0,0 +1 @@
+{ callPackage }: callPackage ../../ne/netmaker/package.nix { withServer = false; withCLI = true; }
```
This should be in all-packages. See the readme in the by-name directory

SuperSandro2000 commented on nixos/modules/services/networking/netmaker/netmaker.nix:
```diff
@@ -0,0 +1,787 @@
+
+        You might want to create following DNS records at your root DNS zone registrar before applying the module:
+
+            api.<YOUR-DOMAIN>.         1       IN      A       <YOUR-SERVER-IP>
+            broker.<YOUR-DOMAIN>.      1       IN      A       <YOUR-SERVER-IP>
+            dashboard.<YOUR-DOMAIN>.   1       IN      A       <YOUR-SERVER-IP>
+            ns.<YOUR-DOMAIN>.          1       IN      A       <YOUR-SERVER-IP>
```
Why not CNAMEs? Do they also work?

SuperSandro2000 commented on nixos/modules/services/networking/netmaker/netmaker.nix:
```diff
@@ -0,0 +1,787 @@
+    email = lib.mkOption {
+      type = with lib.types; str;
+      description = lib.mdDoc ''
+        Email address to (currently) use for certificato registration.
```
```suggestion
        Email address to use for certificate registration.
```

SuperSandro2000 commented on nixos/modules/services/networking/netmaker/netmaker.nix:
```diff
@@ -0,0 +1,787 @@
+
+    package = lib.mkOption {
+      type = with lib.types; package;
+      default = if cfg.installType == "ce" then pkgs.netmaker else pkgs.netmaker-pro;
```
We should move this into config and use the standard mkPackageOption helper.

SuperSandro2000 commented on nixos/modules/services/networking/netmaker/netmaker.nix:
```diff
@@ -0,0 +1,787 @@
+      '';
+    };
+
+    nmctl.package = lib.mkOption {
```
Please use mkPackageOption

SuperSandro2000 commented on nixos/modules/services/networking/netmaker/netmaker.nix:
```diff
@@ -0,0 +1,787 @@
+    };
+
+    internalListenAddress = lib.mkOption {
+      type = with lib.types; str;
```
```suggestion
      type = lib.types.str;
```
we shouldn't use with if it is not necessary

SuperSandro2000 commented on nixos/modules/services/networking/netmaker/netmaker.nix:
```diff
@@ -0,0 +1,787 @@
+        Content of Netmaker configuration file.
+
+        see for more information:
+        - https://github.com/gravitl/netmaker/blob/dc8f9b1bc74d262669bf96fab5f0e545f5906432/config/config.go
```
```suggestion
        <https://github.com/gravitl/netmaker/blob/dc8f9b1bc74d262669bf96fab5f0e545f5906432/config/config.go>
```
should we code the version of the package into the url instead?

SuperSandro2000 commented on nixos/modules/services/networking/netmaker/netmaker.nix:
```diff
@@ -0,0 +1,787 @@
+      ];
+      default = "ce";
+      description = lib.mdDoc ''
+        Type of the installation: `ce` for Community and `pro` for Professional.
+
+        `pro` requires additional configuration not covered by the module, for reference see the [code search](https://github.com/search?q=repo%3Agravitl%2Fnetmaker+%22%24INSTALL_TYPE%22&type=code)
```
If the option is broken, then we should rather drop it.

SuperSandro2000 commented on nixos/modules/services/networking/netmaker/netmaker.nix:
```diff
@@ -0,0 +1,787 @@
+        Netmaker data directory, currently simply a working directory the server is running in.
+        It contains following known directories and files:
+        - `data/netmaker.db` - sqlite database file
+        - `data/netmaker.log.YYYY-MM-DD` - Netmaker server log files
```
log files really belong under /var/log/netmaker

SuperSandro2000 commented on nixos/modules/services/networking/netmaker/netmaker.nix:
```diff
@@ -0,0 +1,787 @@
+        A configuration directory to (currently) put NixOS module related configuration files in.
+
+        Following configuration files are used by default:
+        - `''${config.services.netmaker.configDir}/env` - user editable environment variables
+        - `''${config.services.netmaker.configDir}/env.generated` - generated environment variables
```
editable data belongs under /var/lib. /etc/ only really holds hardcoded configs which we cannot really pass directly to the program or we need a static path for other reasons.

SuperSandro2000 commented on nixos/modules/services/networking/netmaker/netmaker.nix:
```diff
@@ -0,0 +1,787 @@
+      type = with lib.types; bool;
+      default = true;
+      description = lib.mdDoc ''
+        Should Netmaker (web) UI/Dashboard be enabled?
```
```suggestion
        Whether to enable the Netmaker Dashboard.
```

SuperSandro2000 commented on nixos/modules/services/networking/netmaker/netmaker.nix:
```diff
@@ -0,0 +1,787 @@
+      type = with lib.types; str;
+      default = "";
+      description = lib.mdDoc ''
+        A piece configuration related to SaaS build of the UI.
```
I am confused what this is suppose to mean.

SuperSandro2000 commented on nixos/modules/services/networking/netmaker/netmaker.nix:
```diff
@@ -0,0 +1,787 @@
+      type = with lib.types; str;
+      default = "https://account.staging.netmaker.io";
+      description = lib.mdDoc ''
+        A piece configuration related to SaaS build of the UI.
```
This is an URL, right?

SuperSandro2000 commented on nixos/modules/services/networking/netmaker/netmaker.nix:
```diff
@@ -0,0 +1,787 @@
+        default = 10;
+
+        description = lib.mdDoc ''
+          Number of Netmaker networks to open up firewall for:
+          > Netmaker needs one port per network, starting with 51821,
+          > so open up a range depending on the number of networks
+          > you plan on having. For instance, 51821-51830.
```
```suggestion
          Number of Netmaker networks to open up firewall for.

          Netmaker needs one port per network, starting with 51821,
          so open up a range depending on the number of networks you plan on having.
          For instance, 51821-51830.
```

SuperSandro2000 commented on nixos/modules/services/networking/netmaker/netmaker.nix:
```diff
@@ -0,0 +1,787 @@
+
+    webserver.type = lib.mkOption {
+      type = with lib.types; enum [
+        "caddy"
```
Why not nginx? That's the default webserver on nixos.

SuperSandro2000 commented on nixos/modules/services/networking/netmaker/netmaker.nix:
```diff
@@ -0,0 +1,787 @@
+      '';
+    };
+
+    webserver.openFirewall = lib.mkOption {
```
This should be part of the webserver.

SuperSandro2000 commented on nixos/modules/services/networking/netmaker/netmaker.nix:
```diff
@@ -0,0 +1,787 @@
+    };
+    coredns.defaults.forwards = lib.mkOption {
+      type = with lib.types; listOf str;
+      default = [
+        "8.8.8.8"
+        "8.8.4.4"
```
This should default to config.networking.nameservers

SuperSandro2000 commented on nixos/modules/services/networking/netmaker/netmaker.nix:
```diff
@@ -0,0 +1,787 @@
+      }];
+    }
+    {
+      systemd.tmpfiles.rules = [
```
Please use the new syntax

SuperSandro2000 commented on nixos/modules/services/networking/netmaker/netmaker.nix:
```diff
@@ -0,0 +1,787 @@
+        (lib.mkAfter cfg.environment)
+      ];
+
+      services.netmaker.config = {
+        server.apiconn = "${cfg.api.domain}:443";
+        server.apihost = cfg.api.domain;
```
```suggestion
      services.netmaker.config.server = {
        apiconn = "${cfg.api.domain}:443";
        apihost = cfg.api.domain;
```
we can save lots of repetition by moving server

nazarewk replied:
yeah, but makes it a lot harder to track down where the non-NixOS configuration starts and how it maps to the Netmaker configuration. How about something like this?

```
services.netmaker.config = {
	server = {
		....
	};
};
```

SuperSandro2000 commented on nixos/modules/services/networking/netmaker/netmaker.nix:
```diff
@@ -0,0 +1,787 @@
+        serviceConfig.WorkingDirectory = cfg.dataDir;
+        serviceConfig.ExecStart = "${lib.getExe cfg.package} -c ${cfg.config}";
+        serviceConfig.ExecStartPost = pkgs.writeShellScript "configure-nmctl" ''
+          export HOME=/root
```
Should the service really dump it's files into the root users' home?

nazarewk replied:
I don't like it either, but people might expect `nmctl` to work out of the box, because it is part of the default non-NixOS installation https://github.com/gravitl/netmaker/blob/b9563fbff45dcd1fef459fc23a79b7ea6dea7cbe/scripts/nm-quick.sh#L172-L193

SuperSandro2000 commented on nixos/modules/services/networking/netmaker/netmaker.nix:
```diff
@@ -0,0 +1,787 @@
+              jq -r "$@" <<<"$info"
+            }
+
+            export MQTTUI_PASSWORD="''${MQTTUI_PASSWORD:-"$(get '.MQPassword')"}"
+            export MQTTUI_USERNAME="''${MQTTUI_USERNAME:-"$(get '.MQUserName')"}"
+            export MQTTUI_BROKER="''${MQTTUI_BROKER:-"$(get '.Broker')"}"
```
We don't have this information in nix code, do we?

nazarewk replied:
we have the username, but not the password which gets generated on the first run with configuration script so can as well pull the whole thing from the server.

nazarewk replied:
I considered sourcing the configuration files, but then it would include a lot more collateral envvars. I used it to debug actual `netclient` permissions so getting server config the same way as `netclient` made the most sense.

SuperSandro2000 commented on nixos/modules/services/networking/netmaker/netmaker.nix:
```diff
@@ -0,0 +1,787 @@
+      networking.firewall.allowedTCPPorts = lib.optionals cfg.webserver.openFirewall [ 80 443 ];
+      systemd.services.caddy.serviceConfig.AmbientCapabilities = [ "CAP_NET_BIND_SERVICE" ];
+
+      systemd.services.netmaker.after = [ "caddy.service" ];
+      systemd.services.netmaker.requires = [ "caddy.service" ];
```
```suggestion
      systemd.services.netmaker = {
        after = [ "caddy.service" ];
      	requires = [ "caddy.service" ];
      }:
```

SuperSandro2000 commented on nixos/modules/services/networking/netmaker/netmaker.nix:
```diff
@@ -0,0 +1,787 @@
+        }
+      '';
+    })
+    (lib.mkIf (cfg.db.type == "sqlite") { })
```
```suggestion
```
that's currently a noop

2xsaiko commented on pkgs/top-level/all-packages.nix:
```diff
@@ -33988,8 +33988,7 @@ with pkgs;
 
-  netmaker = callPackage ../applications/networking/netmaker {subPackages = ["."];};
-  netmaker-full = callPackage ../applications/networking/netmaker { };
+  netmaker-full = callPackage ../by-name/ne/netmaker/package.nix { withServer = true; withCLI = true; };
```
```suggestion
  netmaker-full = netmaker.override { withServer = true; withCLI = true; };
```

It's probably better to use override instead of re-importing the package, I think. (Same for the other cases of this.)

nazarewk replied:
yeah, that `by-name` validation seems to be a lot happier about override all around

nazarewk commented on nixos/modules/services/networking/netmaker/netmaker-generate-config.sh:
```diff
@@ -0,0 +1,81 @@
+#!/usr/bin/env bash
```
How else would I generate 2 passwords without putting them into the store?

nazarewk commented on nixos/modules/services/networking/netmaker/netclient.nix:
```diff
@@ -0,0 +1,76 @@
+
+    firewall.trusted = lib.mkOption {
+      type = with lib.types; bool;
+      internal = true;
+      readOnly = true;
```
right, I initially made is as non-configurable, but doesn't make sense anymore

nazarewk commented on nixos/modules/services/networking/netmaker/netmaker-generate-config.sh:
```diff
@@ -0,0 +1,81 @@
+  fi
+}
+
+source_envs() {
+  local envFiles
+  mapfile -t envFiles < <(systemctl cat netmaker | grep EnvironmentFile | sed 's/^.*=//g')
+  set -a
+  for file in "${envFiles[@]}"; do
+    if test -e "${file}"; then
+      # shellcheck disable=SC1090
+      . "${file}"
+    fi
+  done
+  set +a
+}
```
makes sense

nazarewk commented on pkgs/by-name/ne/netclient/package.nix:
```diff
@@ -1,34 +1,69 @@
 }:
-
 buildGoModule rec {
   pname = "netclient";
-  version = "0.22.0";
 
+  version = "unstable-2024-01-26";
```
this is a result of `nix-update-script` (with the amount of code fixes applied and tests on a live system I'd go crazy updating it manually).

nazarewk committed:
nixos/netmaker: init

nazarewk committed:
nixos/netclient: add missing configs, move to common netmaker directory

nazarewk commented on nixos/modules/services/networking/netmaker/netclient.nix:
```diff
@@ -0,0 +1,76 @@
+{
+  meta.maintainers = with lib.maintainers; [ wexder ];
+
+  options.services.netclient = {
```
Are we sure we want to rename it? It was `servicet.netclient` for 3 months already https://github.com/NixOS/nixpkgs/commits/nixos-23.11/nixos/modules/services/networking/netclient.nix

nazarewk committed:
netmaker-full: remove package

Seems like the whole suite was built simultaneously from a monorepo.
Now packages are split over 3 repositories and 5 (incl. pro) packages.

nazarewk commented:
I decided to drop `netmaker-full` completely as I'm not sure what was the original intent of bundling it. I can just guess `netclient` was built from `netmaker` repo and made sense to bundle together, but now `netclient` has it's own repo and the bundled CLI is `nmctl` which is better off as an explicit package.

wexder commented:
> I decided to drop `netmaker-full` completely as I'm not sure what was the original intent of bundling it. I can just guess `netclient` was built from `netmaker` repo and made sense to bundle together, but now `netclient` has it's own repo and the bundled CLI is `nmctl` which is better off as an explicit package.

Yes `netmaker` and `netclient` were originally build from one repository. I don't think we need `netmaker-full` anymore

nazarewk committed:
nixos/netmaker: add `services.netmaker.coredns.defaults.bind`

allows limiting default CoreDNS listen addresses.

2xsaiko commented on pkgs/by-name/ne/netclient/package.nix:
```diff
@@ -1,34 +1,61 @@
+  version = "unstable-2024-01-26";
 
   src = fetchFromGitHub {
-    owner = "gravitl";
+    # TODO: switch to upstream after https://github.com/gravitl/netclient/pull/682
+    #owner = "gravitl";
+    owner = "nazarewk";
```
Might be better to have this PR as patches and keep the upstream repo? Not sure what's generally preferred but that's what I usually see.

nazarewk replied:
can do that at this point (got everything as much to work as possible), but it's extremely tedious to update whenever code change is required.

nazarewk commented:
After a few rounds of testing all the functionality I learned that free self-hosted Netmaker version dropped support for peers to talk to each other from behind NAT in favor of a paid feature (failovers) in https://github.com/gravitl/netmaker/releases/tag/v0.22.0 , making it effectively useless for free/personal use.

I don't think I have much interest in continuing this PR apart from getting experience on merging stuff into `nixpkgs`

AndersonTorres commented on pkgs/top-level/aliases.nix:
```diff
@@ -711,6 +711,10 @@ mapAliases ({
   ''; # Added 2023-10-13
   nextcloud25Packages = throw "Nextcloud25 is EOL!"; # Added 2023-10-13
   nagiosPluginsOfficial = monitoring-plugins;
+  netmaker-full = throw ''
+    Netmaker is not built from a monorepo for a while already, so *-full is not justified anymore.
+    Use those instead: netmaker, nmctl, netclient, netmaker-ui.
+  '';
```
```suggestion
  netmaker-full = throw ''netmaker-full was removed. Use the individual packages netmaker, nmctl, netclient, netmaker-ui instead.
  '';
```

AndersonTorres commented on pkgs/by-name/ne/netmaker-pro/package.nix:
```diff
@@ -0,0 +1 @@
+{ netmaker }: netmaker.override { isPro = true; }
```
I don't believe it is a good idea to "abuse" by-name hierarchy this way. 
It is cleaner to put this as an alias on top-level/all-packages.nix instead.

nazarewk replied:
I was not possible for me due to those errors:
```
pkgs.netmaker-pro: This is a new top-level package of the form `callPackage ./pkgs/by-name/ne/netmaker/package.nix { }`. Please define it in pkgs/by-name/ne/netmaker-pro/package.nix instead. See `pkgs/by-name/README.md` for more details. Since the second `callPackage` argument is not `{ }`, the manual `callPackage` (e.g. in `pkgs/top-level/all-packages.nix`) is still needed.
pkgs.nmctl: This is a new top-level package of the form `callPackage ./pkgs/by-name/ne/netmaker/package.nix { }`. Please define it in pkgs/by-name/nm/nmctl/package.nix instead. See `pkgs/by-name/README.md` for more details. Since the second `callPackage` argument is not `{ }`, the manual `callPackage` (e.g. in `pkgs/top-level/all-packages.nix`) is still needed.
```

nazarewk replied:
actually nevermind, I think I used it wrongly

nazarewk commented:
@wexder would you be up for continuing this pull request?

wexder commented:
> @wexder would you be up for continuing this pull request?

@nazarewk I think we should finish this PR, but I'm having hard time finding motivation. I've dropped netmaker in my home lab, because it seems most of the features are being locked behind pro license.

nazarewk commented:
Sad as it is I think the best course is to close this even though much work was done, because:
- I guess there is a lot more work to do to get the licensed installation to work,
- it is not usable at all unless you have all peers exposed directly to internet on a public IP (which defeats the purpose of VPN)
- there is no point putting something that doesn't work into nixpkgs

If anybody wants to follow through with it, feel free to reuse the code and finish in another PR.
